src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -2548,13 +2588,58 @@ public boolean validEncodingQuery(DynamicObject string) {
     }
 
     @CoreMethod(names = "capitalize!", raiseIfFrozenSelf = true)
+    @ImportStatic(StringGuards.class)
     public abstract static class CapitalizeBangNode extends CoreMethodArrayArgumentsNode {
 
         @Child private RopeNodes.GetCodePointNode getCodePointNode = RopeNodes.GetCodePointNode.create();
         @Child private RopeNodes.MakeLeafRopeNode makeLeafRopeNode = RopeNodes.MakeLeafRopeNode.create();
 
-        @Specialization
+        @Specialization(guards = "isSingleByteOptimizable(string)")
+        public DynamicObject capitalizeSingleByte(DynamicObject string,
+                                                  @Cached("createUpperToLower()") InvertAsciiCaseBytesNode invertAsciiCaseNode,
+                                                  @Cached("create()") RopeNodes.BytesNode bytesNode,
+                                                  @Cached("createBinaryProfile()") ConditionProfile emptyStringProfile,
+                                                  @Cached("createBinaryProfile()") ConditionProfile firstCharIsLowerProfile,
+                                                  @Cached("createBinaryProfile()") ConditionProfile otherCharsAlreadyLowerProfile,
+                                                  @Cached("createBinaryProfile()") ConditionProfile mustCapitalizeFirstCharProfile) {
+            final Rope rope = rope(string);
+
+            if (emptyStringProfile.profile(rope.isEmpty())) {
+                return nil();
+            }
+
+            final byte[] sourceBytes = bytesNode.execute(rope);
+            final byte[] finalBytes;
+
+            final byte[] processedBytes = invertAsciiCaseNode.executeInvert(sourceBytes, 1);
+
+            if (otherCharsAlreadyLowerProfile.profile(processedBytes == null)) {
+                // Bytes 1..N are either not letters or already lowercased. Time to check the first byte.
+
+                if (firstCharIsLowerProfile.profile(StringSupport.isAsciiLowercase(sourceBytes[0]))) {
+                    // The first char requires capitalization, but the remaining bytes in the original string are
+                    // already properly cased.
+                    finalBytes = sourceBytes.clone();
```

Comment:
Sounds like a good opportunity to use a ConcatRope + SubstringRope to avoid copying :)

Reply:
I plan to take another pass over this to make it more rope friendly. If the original string had a leaf rope, blowing it up to a concat rope may not be ideal.